File: ruby/acronym/acronym.rb
class Acronym

  VERSION = 1

  def self.abbreviate(phrase)
    new_phrase = extract_acronym(phrase).gsub(/\W/, " ")
    new_phrase.chars.select.with_index do |letter, index|
      letter.upcase == letter || new_phrase[index - 1] == " "
    end.delete_if { |character| character == " " }
       .map { |letter| letter.upcase }.join
  end

  def self.extract_acronym(phrase)
    phrase.include?(":") ? phrase.split(":")[0] : phrase
  end
end

File: python/acronym/acronym.py
import re

def abbreviate(phrase):
    phrase = re.sub("[\W]", " ", phrase)
    acronym = ""
    for index, letter in enumerate(phrase):
        if uppercase_or_leading_letter(phrase, index, letter) and not_space(letter) and previous_letter_not_capitalized(phrase, index, letter):
            acronym += letter
    return acronym.upper()

def uppercase_or_leading_letter(phrase, index, letter):
    return letter == letter.upper() or phrase[index - 1] == " "

def not_space(letter):
    return letter != " "

def previous_letter_not_capitalized(phrase, index, letter):
    return phrase[index - 1] == " " or phrase[index - 1] != phrase[index - 1].upper()
